## Capstone Project

### Modeling Notebook - `DecisionTrees`  `COMBINED DATASET TREE` from 10/24/19

#### Importing Libraries

In [1]:
%matplotlib inline

# general libraries
import re
import string
import sys
import os
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# importing date libraries
import datetime as dt
import dateutil.parser as dparser

# scikit-learn libraries for preprocessing
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelBinarizer
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.preprocessing import FunctionTransformer
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

# scikit-learn libraries for constructing pipelines
from sklearn.pipeline import FeatureUnion, Pipeline
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import FunctionTransformer
from sklearn.base import BaseEstimator, TransformerMixin

# scikit-learn libraries for clustering and dimensionality reduction
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.cluster import DBSCAN
from scipy.cluster.hierarchy import dendrogram, linkage, fcluster
from sklearn.mixture import GaussianMixture

# scikit-learn libraries for evaluation
from sklearn import metrics

# scikit-learn libraries for feature selection
from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2, f_classif
from sklearn.feature_selection import SelectPercentile
from sklearn.feature_selection import RFECV

# scikit-learn libraries for learning
from sklearn.model_selection import KFold, StratifiedKFold, GridSearchCV, cross_validate, cross_val_score
from sklearn.model_selection import validation_curve
from sklearn.model_selection import RandomizedSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

# saving models
import pickle
from sklearn.externals import joblib

# setting pandas display options
pd.set_option("display.max_columns", 999)
pd.set_option("display.max_rows", 10000)
pd.set_option('display.max_colwidth', 100)
pd.set_option('precision', 5)
pd.options.mode.chained_assignment = None

/Users/nate_velarde/anaconda/envs/py3jp/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)


#### Directory/File Structure

In [2]:
sys.version

'3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) \n[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]'

In [3]:
print ('Running pandas version:', pd.__version__)
print ('Running numpy version:', np.__version__)
print ('Running sklearn version:', sklearn.__version__)

Running pandas version: 0.25.1
Running numpy version: 1.14.2
Running sklearn version: 0.21.3


In [4]:
os.getcwd()

'/Users/nate_velarde/Documents/UC_Berkeley/Courses/W210_Capstone/stroke_project/sandbox/notebooks'

In [5]:
os.chdir('../data')

In [6]:
sorted(os.listdir())

['.DS_Store',
 '273_vs_281_null_count_by_feature.csv',
 '273_vs_281_null_count_by_feature.xlsm',
 'Capstone - Complication list - complete.xlsx',
 'Capstone - STS risk factor list.xlsx',
 'Capstone_Fall_Shannon_Sept2019_request.csv',
 'PREOP_dataset_10_24.pkl',
 'PREOP_dataset_TREE_10_24.pkl',
 'X_A_DREF.pkl',
 'X_A_DREF_TREE_SKLEARN.pkl',
 'X_PREOP_10_24.pkl',
 'X_PREOP_TREE_10_24.pkl',
 'X_dev_A_DREF.pkl',
 'X_dev_A_DREF_TREE_SKLEARN.pkl',
 'X_dev_PREOP_10_24.pkl',
 'X_dev_PREOP_TREE_10_24.pkl',
 'X_test_A_DREF.pkl',
 'X_test_A_DREF_TREE_SKLEARN.pkl',
 'X_test_PREOP_10_24.pkl',
 'X_test_PREOP_TREE_10_24.pkl',
 'X_train_A_DREF.pkl',
 'X_train_A_DREF_TREE_SKLEARN.pkl',
 'X_train_PREOP_10_24.pkl',
 'X_train_PREOP_TREE_10_24.pkl',
 'capstone_STS_risk_factor_features.xlsx',
 'capstone_cleaned_data.csv',
 'capstone_data-version-2.xlsx',
 'capstone_data.xlsx',
 'capstone_data_binarized_outcome.pkl',
 'capstone_data_binarized_outcome.xlsx',
 'capstone_data_binarized_outcome_compressed.pkl',


#### Loading Datasets

#### `X_train`, `y_train`
- designation of `_all` denotes complete feature set

In [7]:
X_train_all = pd.read_pickle('X_train_PREOP_TREE_10_24.pkl')
y_train = pd.read_pickle('y_train_PREOP_TREE_10_24.pkl')

In [8]:
X_train_all.shape, y_train.shape

((34192, 73), (34192,))

#### `X_dev`, `y_dev`
- designation of `_all` denotes complete feature set

In [9]:
X_dev_all = pd.read_pickle('X_dev_PREOP_TREE_10_24.pkl')
y_dev = pd.read_pickle('y_dev_PREOP_TREE_10_24.pkl')

In [10]:
X_dev_all.shape, y_dev.shape

((4274, 73), (4274,))

#### `X_test`, `y_test`
- designation of `_all` denotes complete feature set

In [11]:
X_test_all = pd.read_pickle('X_dev_PREOP_TREE_10_24.pkl')
y_test = pd.read_pickle('y_dev_PREOP_TREE_10_24.pkl')

In [12]:
X_test_all.shape, y_dev.shape

((4274, 73), (4274,))

- validating row count for `COMBINED DATASET TREE` from 10/24/19 - `42,740` total observations

In [13]:
42740 - X_train_all.shape[0] - X_dev_all.shape[0] - X_test_all.shape[0]

0

- last look at the data (`X_train_all`) before modeling

In [14]:
X_train_all.head()

age  heightcm  weightkg      bmi      hct  creatlst  totalbumin  \
0 -2.12457   0.11594   0.32942  0.15700 -2.65268  -0.35221    -2.27155   
1  1.08243  -0.82771  -0.95497 -0.39563 -0.10617  -0.13691     0.04597   
2 -0.20037   1.53140   1.56497  0.35976  0.59504  -0.25533    -0.34028   
3  0.44103  -0.31888  -0.21754 -0.04732  0.59504  -0.25533     0.04597   
4 -0.75014  -1.05900   0.19268  0.56118 -2.17290  -0.36298     0.04597   

    a1clvl  meldscr     hdef    pasys  surgdt_month  surgdt_DayOfWeek  \
0 -1.62927  2.47597  0.98148  3.58585            10                 2   
1  0.25206 -0.29493  0.33103  0.49141            10                 0   
2  0.32442 -0.67978  0.57495 -0.12748            10                 1   
3 -0.76096 -0.29493  0.16842 -0.64322            11                 3   
4  0.39678 -0.29493  0.57495 -0.65353             6                 4   

  surgdt_PartOfMonth  gender  racecaucasian  raceblack  raceasian  \
0                  2     1.0            1.0        0.0        0.0   
1                  3     0.0            1.0        0.0        0.0   
2                  1     1.0            1.0        0.0        0.0   
3                  3     1.0            1.0        0.0        0.0   
4                  3     0.0            1.0        0.0        0.0   

   racenativeam  racnativepacific  ethnicity  diabetes  dyslip  dialysis  \
0           0.0               0.0        0.0       1.0     1.0       0.0   
1           0.0               0.0        0.0       1.0     1.0       0.0   
2           0.0               0.0        0.0       1.0     1.0       0.0   
3           0.0               0.0        0.0       0.0     1.0       0.0   
4           0.0               0.0        0.0       1.0     1.0       0.0   

   hypertn  infendo  slpapn  liverdis  immsupp  mediastrad  cancer  pvd  \
0        1      1.0     0.0       0.0        0         0.0     0.0  0.0   
1        0      0.0     0.0       0.0        0         0.0     1.0  1.0   
2        1      0.0     0.0       0.0        1         0.0     0.0  0.0   
3        0      0.0     0.0       0.0        0         0.0     0.0  0.0   
4        1      0.0     0.0       0.0        0         0.0     0.0  0.0   

   syncope  unrespstat  cvd  cva  cvdtia  cvdpcarsurg  hitanti  prcvint  \
0      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   
1      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   
2      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   
3      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   
4      0.0         0.0  1.0  0.0     0.0          0.0      0.0      0.0   

   prcab  prvalve  chf  priorhf  arrhyafib  medinotr  hdefd  vdaort  vdstena  \
0    0.0      0.0  1.0      1.0        0.0       0.0    1.0     1.0      0.0   
1    0.0      0.0  1.0      1.0        0.0       0.0    1.0     0.0      0.0   
2    0.0      0.0  0.0      0.0        0.0       0.0    1.0     0.0      0.0   
3    0.0      0.0  0.0      0.0        0.0       0.0    1.0     0.0      0.0   
4    0.0      0.0  0.0      0.0        0.0       0.0    1.0     0.0      0.0   

   vdstenm  diabctrl  infendty  Tobacco_Combined  chrlungd  hmo2  ivdrugab  \
0      0.0       0.0       1.0               1.0       0.0   0.0       0.0   
1      0.0       0.0       0.0               0.0       1.0   0.0       0.0   
2      0.0       0.0       0.0               0.0       0.0   0.0       0.0   
3      0.0       0.0       0.0               0.0       0.0   0.0       0.0   
4      0.0       1.0       0.0               0.0       0.0   0.0       0.0   

   alcohol  carshock24  resusc24  medasa  medaplt5days  medlipid  numdisv  \
0      0.0         0.0       0.0     0.0             0       0.0      0.0   
1      0.0         0.0       0.0     1.0             0       1.0      1.0   
2      0.0         0.0       0.0     1.0             0       1.0      1.0   
3      0.0         0.0       0.0     1.0             0       1.0      1.0   
4      0.0    

In [15]:
X_train_all.tail()

age  heightcm  weightkg      bmi      hct  creatlst  totalbumin  \
34187 -0.29200   0.58776   0.78848  0.26686  0.04145  -0.25533     0.04597   
34188 -1.11666   0.60626  -0.14917 -0.31935 -0.51214  -0.47063    -2.07842   
34189  0.62428  -1.05900   0.17314  0.54625 -0.32761  -0.07232    -0.34028   
34190  0.99080  -1.52157   0.75918  1.25109 -0.19844  -0.47063     0.62535   
34191 -0.20037   0.60626   0.71034  0.21143 -0.69667  -0.04003    -0.14715   

        a1clvl  meldscr     hdef    pasys  surgdt_month  surgdt_DayOfWeek  \
34187  0.83093 -0.29493  0.00581  0.56361             1                 3   
34188  0.61386  0.57972 -3.16512  1.62604             7                 4   
34189 -1.05040 -0.45237  1.38800 -0.12748             3                 0   
34190  1.04801 -0.67978  0.57495  2.03863            10                 3   
34191  3.14642 -0.36491  0.81886 -0.95266            12                 1   

      surgdt_PartOfMonth  gender  racecaucasian  raceblack  raceasian  \
34187                  1     1.0            1.0        0.0        0.0   
34188                  3     1.0            1.0        0.0        0.0   
34189                  3     0.0            1.0        0.0        0.0   
34190                  3     0.0            0.0        1.0        0.0   
34191                  3     1.0            1.0        0.0        0.0   

       racenativeam  racnativepacific  ethnicity  diabetes  dyslip  dialysis  \
34187           0.0               0.0        0.0       1.0     1.0       0.0   
34188           0.0               0.0        0.0       1.0     1.0       0.0   
34189           0.0               0.0        0.0       0.0     1.0       0.0   
34190           0.0               0.0        0.0       1.0     0.0       0.0   
34191           0.0               0.0        0.0       1.0     1.0       0.0   

       hypertn  infendo  slpapn  liverdis  immsupp  mediastrad  cancer  pvd  \
34187        1      0.0     0.0       0.0        1         0.0     0.0  0.0   
34188        1      0.0     0.0       0.0        0         0.0     0.0  0.0   
34189        1      0.0     0.0       0.0        0         0.0     0.0  0.0   
34190        1      0.0     1.0       0.0        0         0.0     0.0  0.0   
34191        1      0.0     0.0       0.0        0         0.0     0.0  0.0   

       syncope  unrespstat  cvd  cva  cvdtia  cvdpcarsurg  hitanti  prcvint  \
34187      0.0         0.0  0.0  0.0     0.0          0.0      0.0      1.0   
34188      0.0         0.0  0.0  0.0     0.0          0.0      0.0      1.0   
34189      0.0         0.0  1.0  0.0     0.0          0.0      0.0      0.0   
34190      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   
34191      0.0         0.0  0.0  0.0     0.0          0.0      0.0      0.0   

       prcab  prvalve  chf  priorhf  arrhyafib  medinotr  hdefd  vdaort  \
34187    0.0      0.0  0.0      0.0        0.0       0.0    1.0     1.0   
34188    0.0      0.0  1.0      0.0        0.0       1.0    1.0     0.0   
34189    0.0      0.0  1.0      0.0        0.0       0.0    1.0     1.0   
34190    0.0      0.0  0.0      0.0        0.0       0.0    1.0     1.0   
34191    0.0      0.0  0.0      0.0        0.0       0.0    1.0     1.0   

       vdstena  vdstenm  diabctrl  infendty  Tobacco_Combined  chrlungd  hmo2  \
34187      0.0      0.0       0.0       0.0               0.0       0.0   0.0   
34188      0.0      0.0       0.0       0.0               0.0       0.0   0.0   
34189      1.0      0.0       0.0       0.0               0.0       0.0   0.0   
34190      1.0      1.0       1.0       0.0               1.0       0.0   0.0   
34191      1.0      0.0       0.0       0.0               0.0       0.0   0.0   

       ivdrugab  alcohol  carshock24  resusc24  medasa  medaplt5days  \
34187       0.0      0.0         0.0       0.0     1.0             0   
34188       0.0      0.0         0.0       0.0     1.0             0   
34189       0.0      0.0         0.0       0.0     1.0 

## Random Forest Classifier
- see Albon 14.4 pages 238-240

- instantiating a `RandomForestClassifier()` object

In [16]:
rf_clf = RandomForestClassifier(class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

#### Default `RandomForestClassifier()` Settings
- `n_estimators`: `10` -- number of trees in the forest
- `criterion`: `gini` -- function to measure the quality of split -- could also set for `entropy` 
- `max_depth`: `None` -- the maximum depth of the tree -- if `None`, then nodes are expanded until all leaves are pure or until all leaves contain less than min_samples_split samples 
- `max_features`: `Auto` -- `max_features=sqrt(n_features)` -- can put in a `float` to represent the number of features used -- per `Muller and Guido` page 88, increasing `max_features` might sometimes improve performance
- `min_samples_split`: `2` -- minimum number of samples required to split an internal node

- running out of the box with default settings only takes a few seconds

- creating a `StratifiedKFold` cross validation

In [17]:
skf = StratifiedKFold(n_splits=5,
                      shuffle=True,
                      random_state=0)

- defining `scoring_metrics`

In [18]:
scoring_metrics = ['accuracy',
                   'f1',
                   'f1_macro',
                   'f1_weighted',
                   'precision',
                   'precision_macro',
                   'precision_weighted',
                   'recall',
                   'recall_macro',
                   'recall_weighted',
                   'roc_auc']

- conduct `StratifiedKFold` cross-validation

In [19]:
skf_results = cross_validate(rf_clf, # model
                             X_train_all, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores                     

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   11.4s finished


In [20]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.98289  0.98289  0.98304  0.98304  0.98304   
train_accuracy            0.99580  0.99543  0.99565  0.99525  0.99481   
test_f1                   0.00000  0.00000  0.00000  0.00000  0.00000   
train_f1                  0.85890  0.84472  0.85363  0.83791  0.82025   
test_f1_macro             0.49569  0.49569  0.49572  0.49572  0.49572   
train_f1_macro            0.92838  0.92120  0.92571  0.91775  0.90881   
test_f1_weighted          0.97441  0.97441  0.97463  0.97463  0.97463   
train_f1_weighted         0.99550  0.99508  0.99534  0.99487  0.99435   
test_precision            0.00000  0.00000  0.00000  0.00000  0.00000   
train_precision           1.00000  1.00000  1.00000  1.00000  1.00000   
test_precision_macro      0.49145  0.49145  0.49152  0.49152  0.49152   
train_precision_macro     0.99787  0.99769  0.99780  0.99759  0.99737   
test_precision_weighted   0.96608  0.96608  0.96636  0.96636  0.96636   
train_precision_weighted  0.99581  0.99545  0.99567  0.99527  0.99484   
test_recall               0.00000  0.00000  0.00000  0.00000  0.00000   
train_recall              0.75269  0.73118  0.74464  0.72103  0.69528   
test_recall_macro         0.50000  0.50000  0.50000  0.50000  0.50000   
train_recall_macro        0.87634  0.86559  0.87232  0.86052  0.84764   
test_recall_weighted      0.98289  0.98289  0.98304  0.98304  0.98304   
train_recall_weighted     0.99580  0.99543  0.99565  0.99525  0.99481   
test_roc_auc              0.52308  0.53880  0.52051  0.52142  0.54047   
train_roc_auc             1.00000  0.99999  1.00000  1.00000  0.99999   

                          mean_score  
test_accuracy                0.98298  
train_accuracy               0.99539  
test_f1                      0.00000  
train_f1                     0.84308  
test_f1_macro                0.49571  
train_f1_macro               0.92037  
test_f1_weighted             0.97454  
train_f1_weighted            0.99503  
test_precision               0.00000  
train_precision              1.00000  
test_precision_macro         0.49149  
train_precision_macro        0.99766  
test_precision_weighted      0.96625  
train_precision_weighted     0.99541  
test_recall                  0.00000  
train_recall                 0.72896  
test_recall_macro            0.50000  
train_recall_macro           0.86448  
test_recall_weighted         0.98298  
train_recall_weighted        0.99539  
test_roc_auc                 0.52886  
train_roc_auc                0.99999

#### Key Takeaways
- Model is clearly overfitting
- Can tune `max_depth` parameter to see if that reduces overfitting - pre-prune trees

#### Varying `max_depth`

- instantiating a `RandomForestClassifier()` object

In [21]:
rf_clf = RandomForestClassifier(max_depth=5,
                                class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

In [22]:
skf_results = cross_validate(rf_clf, # model
                             X_train_all, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores                     

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   11.2s finished


In [23]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.74762  0.76956  0.74130  0.75271  0.78078   
train_accuracy            0.75330  0.76639  0.75360  0.75993  0.78449   
test_f1                   0.05476  0.05967  0.05350  0.05794  0.05066   
train_f1                  0.08115  0.07952  0.07823  0.08141  0.08817   
test_f1_macro             0.45457  0.46418  0.45184  0.45781  0.46337   
train_f1_macro            0.46934  0.47287  0.46801  0.47166  0.48299   
test_f1_weighted          0.84069  0.85485  0.83666  0.84411  0.86208   
train_f1_weighted         0.84432  0.85284  0.84451  0.84862  0.86435   
test_precision            0.02926  0.03207  0.02852  0.03097  0.02734   
train_precision           0.04332  0.04261  0.04178  0.04354  0.04751   
test_precision_macro      0.50810  0.50969  0.50777  0.50928  0.50660   
train_precision_macro     0.51758  0.51678  0.51650  0.51754  0.51952   
test_precision_weighted   0.97056  0.97097  0.97076  0.97137  0.96960   
train_precision_weighted  0.97572  0.97482  0.97505  0.97538  0.97544   
test_recall               0.42735  0.42735  0.43103  0.44828  0.34483   
train_recall              0.64086  0.59355  0.61373  0.62446  0.61159   
test_recall_macro         0.59027  0.60143  0.58884  0.60312  0.56657   
train_recall_macro        0.69805  0.68146  0.68488  0.69337  0.69954   
test_recall_weighted      0.74762  0.76956  0.74130  0.75271  0.78078   
train_recall_weighted     0.75330  0.76639  0.75360  0.75993  0.78449   
test_roc_auc              0.66567  0.63947  0.64662  0.66334  0.63110   
train_roc_auc             0.76672  0.76803  0.77043  0.76588  0.77994   

                          mean_score  
test_accuracy                0.75839  
train_accuracy               0.76354  
test_f1                      0.05531  
train_f1                     0.08170  
test_f1_macro                0.45835  
train_f1_macro               0.47297  
test_f1_weighted             0.84768  
train_f1_weighted            0.85093  
test_precision               0.02963  
train_precision              0.04375  
test_precision_macro         0.50829  
train_precision_macro        0.51758  
test_precision_weighted      0.97065  
train_precision_weighted     0.97528  
test_recall                  0.41577  
train_recall                 0.61684  
test_recall_macro            0.59005  
train_recall_macro           0.69146  
test_recall_weighted         0.75839  
train_recall_weighted        0.76354  
test_roc_auc                 0.64924  
train_roc_auc                0.77020

#### Key Takeaways
- Varying `max_depth` significantly reduced overfitting
- We can next try increasing `n_estimators`
- Per `Muller and Guido` in `Introduction to Machine Learning with Python` page 88, "for `n_estimators`, larger is always better as averaging over more trees will yield a more robust ensemble by reducing overfitting"

#### Varying `n_estimators`

- instantiating a `RandomForestClassifier()` object

In [24]:
rf_clf = RandomForestClassifier(n_estimators=100,
                                max_depth=5,
                                class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

In [25]:
skf_results = cross_validate(rf_clf, # model
                             X_train_all, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores              

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   14.6s finished


In [26]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.79003  0.78871  0.77347  0.79365  0.79014   
train_accuracy            0.79721  0.79041  0.78464  0.79791  0.79689   
test_f1                   0.06753  0.05863  0.05144  0.06121  0.05778   
train_f1                  0.09878  0.09702  0.09578  0.10055  0.10068   
test_f1_macro             0.47461  0.46982  0.46141  0.47265  0.46985   
train_f1_macro            0.49227  0.48923  0.48677  0.49336  0.49310   
test_f1_weighted          0.86777  0.86693  0.85747  0.87013  0.86794   
train_f1_weighted         0.87237  0.86811  0.86444  0.87278  0.87214   
test_precision            0.03654  0.03173  0.02769  0.03317  0.03127   
train_precision           0.05343  0.05235  0.05158  0.05440  0.05445   
test_precision_macro      0.51227  0.50923  0.50689  0.51016  0.50901   
train_precision_macro     0.52300  0.52252  0.52218  0.52358  0.52364   
test_precision_weighted   0.97172  0.97038  0.96983  0.97097  0.97053   
train_precision_weighted  0.97660  0.97670  0.97674  0.97677  0.97685   
test_recall               0.44444  0.38462  0.36207  0.39655  0.37931   
train_recall              0.65376  0.66237  0.66953  0.66309  0.66738   
test_recall_macro         0.62024  0.59018  0.57132  0.59853  0.58827   
train_recall_macro        0.72673  0.72749  0.72808  0.73167  0.73326   
test_recall_weighted      0.79003  0.78871  0.77347  0.79365  0.79014   
train_recall_weighted     0.79721  0.79041  0.78464  0.79791  0.79689   
test_roc_auc              0.69596  0.65088  0.65193  0.66621  0.64625   
train_roc_auc             0.81355  0.81074  0.81095  0.80966  0.81539   

                          mean_score  
test_accuracy                0.78720  
train_accuracy               0.79341  
test_f1                      0.05932  
train_f1                     0.09856  
test_f1_macro                0.46967  
train_f1_macro               0.49095  
test_f1_weighted             0.86605  
train_f1_weighted            0.86997  
test_precision               0.03208  
train_precision              0.05324  
test_precision_macro         0.50951  
train_precision_macro        0.52298  
test_precision_weighted      0.97069  
train_precision_weighted     0.97673  
test_recall                  0.39340  
train_recall                 0.66323  
test_recall_macro            0.59371  
train_recall_macro           0.72944  
test_recall_weighted         0.78720  
train_recall_weighted        0.79341  
test_roc_auc                 0.66224  
train_roc_auc                0.81206

#### Key Takeaways
- marginal improvement from increasing `n_estimators`, but it something we can tune using `GridSearchCV`
- let's try changing `criterion` to `entropy`

#### Changing `criterion` to `entropy`

- instantiating a `RandomForestClassifier()` object

In [27]:
rf_clf = RandomForestClassifier(n_estimators=100,
                                max_depth=5,
                                criterion='entropy',
                                class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

In [28]:
skf_results = cross_validate(rf_clf, # model
                             X_train_all, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores              

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   15.4s finished


In [29]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.79003  0.78637  0.76353  0.78795  0.77815   
train_accuracy            0.79191  0.78909  0.77952  0.79517  0.78372   
test_f1                   0.07115  0.06883  0.05050  0.05844  0.05718   
train_f1                  0.09651  0.09591  0.09404  0.09760  0.09430   
test_f1_macro             0.47639  0.47409  0.45772  0.46948  0.46573   
train_f1_macro            0.48946  0.48827  0.48426  0.49104  0.48575   
test_f1_weighted          0.86777  0.86548  0.85113  0.86657  0.86042   
train_f1_weighted         0.86905  0.86728  0.86119  0.87107  0.86386   
test_precision            0.03849  0.03719  0.02710  0.03160  0.03081   
train_precision           0.05210  0.05172  0.05056  0.05276  0.05077   
test_precision_macro      0.51351  0.51275  0.50660  0.50924  0.50886   
train_precision_macro     0.52231  0.52215  0.52166  0.52261  0.52168   
test_precision_weighted   0.97229  0.97203  0.96983  0.97068  0.97068   
train_precision_weighted  0.97653  0.97659  0.97672  0.97645  0.97653   
test_recall               0.47009  0.46154  0.37069  0.38793  0.39655   
train_recall              0.65376  0.65806  0.67167  0.65021  0.66094   
test_recall_macro         0.63284  0.62678  0.57050  0.59139  0.59064   
train_recall_macro        0.72403  0.72471  0.72653  0.72395  0.72340   
test_recall_weighted      0.79003  0.78637  0.76353  0.78795  0.77815   
train_recall_weighted     0.79191  0.78909  0.77952  0.79517  0.78372   
test_roc_auc              0.69617  0.65002  0.64999  0.66294  0.64934   
train_roc_auc             0.81311  0.81015  0.80751  0.80201  0.80947   

                          mean_score  
test_accuracy                0.78121  
train_accuracy               0.78788  
test_f1                      0.06122  
train_f1                     0.09567  
test_f1_macro                0.46868  
train_f1_macro               0.48776  
test_f1_weighted             0.86228  
train_f1_weighted            0.86649  
test_precision               0.03304  
train_precision              0.05158  
test_precision_macro         0.51019  
train_precision_macro        0.52208  
test_precision_weighted      0.97110  
train_precision_weighted     0.97656  
test_recall                  0.41736  
train_recall                 0.65893  
test_recall_macro            0.60243  
train_recall_macro           0.72452  
test_recall_weighted         0.78121  
train_recall_weighted        0.78788  
test_roc_auc                 0.66169  
train_roc_auc                0.80845

#### Key Takeaways
- no material change and also took longer to calculate, so will go back to using default of `gini`

### Hyperparameter Tuning with `RandomizedSearchCV`
- there are lots of other parameters to tune, so let's try to narrow down our possibilities
- this took more than 10 minutes to run, so saved model which will be reloaded in subsequent runs

In [30]:
random_grid = {'n_estimators': [100, 150, 200, 250],          # default is `10`
               'max_features': ['auto', 0.25, 0.50],          # default is square root of `n_features`
               'max_depth': [2, 5, 10, 20, None],             # default is `None`
               'min_samples_split': [2, 5, 10]}               # default is `2`

- instantiating the base model to tune

In [31]:
rf_clf = RandomForestClassifier(class_weight='balanced')

- Random search of parameters, using 3 fold cross validation
- Search across 50 different combinations, and use all available cores

In [32]:
rf_random_model = RandomizedSearchCV(estimator=rf_clf, 
                                     param_distributions=random_grid, 
                                     n_iter=100, 
                                     cv=3,
                                     scoring='roc_auc',
                                     verbose=2, 
                                     random_state=42, 
                                  n_jobs=-1)

- fitting the random search model

In [33]:
#rf_random_model.fit(X_train_all, y_train)

- takes 10 minutes to run

#### Saving `rf_random model`

In [34]:
#joblib.dump(rf_random_model, 'rf_randomized_search.sav')

#### Loading saved `rf_random model`

In [35]:
loaded_rf_random_model = joblib.load('rf_randomized_search.sav')

In [36]:
loaded_rf_random_model.best_params_

{'n_estimators': 100,
 'min_samples_split': 2,
 'max_features': 'auto',
 'max_depth': 2}

- examining results

In [37]:
randomized_search_results = pd.DataFrame(loaded_rf_random_model.cv_results_)

In [38]:
randomized_search_results = randomized_search_results.sort_values(by=['mean_test_score'],
                                                                  ascending=False)

In [39]:
display_cols = ['params',
                'mean_test_score']

In [40]:
randomized_search_results_summary = randomized_search_results.copy()[display_cols].reset_index(drop=True)

In [41]:
randomized_search_results_summary.head(20)

params  \
0    {'n_estimators': 100, 'min_samples_split': 2, 'max_features': 'auto', 'max_depth': 2}   
1   {'n_estimators': 150, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 2}   
2    {'n_estimators': 200, 'min_samples_split': 5, 'max_features': 'auto', 'max_depth': 2}   
3    {'n_estimators': 100, 'min_samples_split': 5, 'max_features': 'auto', 'max_depth': 2}   
4   {'n_estimators': 250, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 2}   
5     {'n_estimators': 200, 'min_samples_split': 10, 'max_features': 0.25, 'max_depth': 2}   
6   {'n_estimators': 200, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 2}   
7   {'n_estimators': 200, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 5}   
8    {'n_estimators': 250, 'min_samples_split': 2, 'max_features': 'auto', 'max_depth': 5}   
9    {'n_estimators': 200, 'min_samples_split': 2, 'max_features': 'auto', 'max_depth': 2}   
10  {'n_estimators': 100, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 5}   
11   {'n_estimators': 200, 'min_samples_split': 2, 'max_features': 'auto', 'max_depth': 5}   
12   {'n_estimators': 100, 'min_samples_split': 2, 'max_features': 'auto', 'max_depth': 5}   
13     {'n_estimators': 250, 'min_samples_split': 2, 'max_features': 0.25, 'max_depth': 2}   
14   {'n_estimators': 200, 'min_samples_split': 5, 'max_features': 'auto', 'max_depth': 5}   
15     {'n_estimators': 100, 'min_samples_split': 2, 'max_features': 0.25, 'max_depth': 2}   
16     {'n_estimators': 200, 'min_samples_split': 5, 'max_features': 0.25, 'max_depth': 2}   
17     {'n_estimators': 250, 'min_samples_split': 5, 'max_features': 0.25, 'max_depth': 2}   
18  {'n_estimators': 150, 'min_samples_split': 10, 'max_features': 'auto', 'max_depth': 5}   
19     {'n_estimators': 100, 'min_samples_split': 5, 'max_features': 0.25, 'max_depth': 2}   

    mean_test_score  
0           0.66698  
1           0.66630  
2           0.66563  
3           0.66499  
4           0.66461  
5           0.66416  
6           0.66375  
7           0.66328  
8           0.66297  
9           0.66276  
10          0.66224  
11          0.66217  
12          0.66195  
13          0.66177  
14          0.66149  
15          0.66148  
16          0.66014  
17          0.66008  
18          0.65973  
19          0.65964

#### Key Takeaways
- clear that `max_depth` is the key parameter - dominates the top 10 results
- will run a `GridSearchCV` to confirm

### Tuning `RandomForestClassifier()` Hyperparameters using `GridSearchCV`
- instantiating a `RandomForestClassifier()` object

In [42]:
rf_clf = RandomForestClassifier(class_weight='balanced',
                                random_state=0)

- defining hyperparamter candidates

In [43]:
hyper_params = {'n_estimators': [200],          
                'max_features': ['auto'],          
                'max_depth': [2, 5],             
                'min_samples_split': [2, 5, 10]}               

- create `GridSearchCV` object

In [44]:
gs = GridSearchCV(rf_clf, # estimator
                  hyper_params, # param_grid
                  scoring=scoring_metrics, # loss functions
                  cv=5, 
                  refit='roc_auc', # best_estimator_ will be based on this scoring metric
                  verbose=2,
                  return_train_score=True,
                  n_jobs=-1)

- fit `GridSearchcV`

In [45]:
rf_gs_model = gs.fit(X_train_all,
                     y_train)

Fitting 5 folds for each of 6 candidates, totalling 30 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  30 out of  30 | elapsed:  1.5min finished


- examining results

In [46]:
gs_results = pd.DataFrame(rf_gs_model.cv_results_)

In [47]:
gs_results = gs_results.sort_values(by=['mean_test_roc_auc'],
                                           ascending=False)

In [48]:
display_cols = ['params',
                'rank_test_roc_auc',
                'mean_test_roc_auc',
                'mean_train_roc_auc']

In [49]:
gs_results_summary = gs_results.copy()[display_cols].reset_index(drop=True)

In [50]:
gs_results_summary

params  \
0   {'max_depth': 2, 'max_features': 'auto', 'min_samples_split': 2, 'n_estimators': 200}   
1   {'max_depth': 2, 'max_features': 'auto', 'min_samples_split': 5, 'n_estimators': 200}   
2  {'max_depth': 2, 'max_features': 'auto', 'min_samples_split': 10, 'n_estimators': 200}   
3  {'max_depth': 5, 'max_features': 'auto', 'min_samples_split': 10, 'n_estimators': 200}   
4   {'max_depth': 5, 'max_features': 'auto', 'min_samples_split': 2, 'n_estimators': 200}   
5   {'max_depth': 5, 'max_features': 'auto', 'min_samples_split': 5, 'n_estimators': 200}   

   rank_test_roc_auc  mean_test_roc_auc  mean_train_roc_auc  
0                  1            0.66622             0.70836  
1                  1            0.66622             0.70836  
2                  1            0.66622             0.70836  
3                  4            0.66293             0.81551  
4                  5            0.66246             0.81561  
5                  6            0.66244             0.81561

### Identifying Important Features in Random Forests
- see Albon 14.6, page 241-243

#### Calculating Feature Importances

In [51]:
importances = rf_gs_model.best_estimator_.feature_importances_

- sort importances in descending order

In [52]:
indicies = np.argsort(importances)[::-1]

- rearrange feature names so they match the sorted feature importances

In [53]:
names = [X_train_all.columns.tolist()[i] for i in indicies]

In [54]:
important_features = pd.DataFrame(list(zip(indicies, 
                                           names, 
                                           importances)),
                                  columns=['index_number', 
                                           'feature_name', 
                                           'feature_importance']).sort_values(by='feature_importance',
                                                                              ascending=False).reset_index(drop=True)

- top features

In [55]:
important_features.head(10)

index_number  feature_name  feature_importance
0             0           age             0.13122
1            11  surgdt_month             0.10837
2             2      weightkg             0.08837
3            34           cvd             0.08445
4             4           hct             0.06592
5            30        cancer             0.05619
6            50      diabctrl             0.05016
7            35           cva             0.04937
8            31           pvd             0.04759
9            18  racenativeam             0.02515

- selecting the top 25 features ranked by feature importance

In [56]:
top_features = important_features['feature_name'].values.tolist()[:25]

### Let's see if reducing the number of features improves results

- instantiating a `RandomForestClassifier()` object

In [57]:
rf_clf = RandomForestClassifier(n_estimators=200,
                                class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

- using only the top features identified above

In [58]:
X_train_top = X_train_all.copy()[top_features]

In [59]:
skf_results = cross_validate(rf_clf, # model
                             X_train_top, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores              

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   33.0s finished


In [60]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.98289  0.98289  0.98304  0.98304  0.98304   
train_accuracy            1.00000  1.00000  1.00000  1.00000  1.00000   
test_f1                   0.00000  0.00000  0.00000  0.00000  0.00000   
train_f1                  1.00000  1.00000  1.00000  1.00000  1.00000   
test_f1_macro             0.49569  0.49569  0.49572  0.49572  0.49572   
train_f1_macro            1.00000  1.00000  1.00000  1.00000  1.00000   
test_f1_weighted          0.97441  0.97441  0.97463  0.97463  0.97463   
train_f1_weighted         1.00000  1.00000  1.00000  1.00000  1.00000   
test_precision            0.00000  0.00000  0.00000  0.00000  0.00000   
train_precision           1.00000  1.00000  1.00000  1.00000  1.00000   
test_precision_macro      0.49145  0.49145  0.49152  0.49152  0.49152   
train_precision_macro     1.00000  1.00000  1.00000  1.00000  1.00000   
test_precision_weighted   0.96608  0.96608  0.96636  0.96636  0.96636   
train_precision_weighted  1.00000  1.00000  1.00000  1.00000  1.00000   
test_recall               0.00000  0.00000  0.00000  0.00000  0.00000   
train_recall              1.00000  1.00000  1.00000  1.00000  1.00000   
test_recall_macro         0.50000  0.50000  0.50000  0.50000  0.50000   
train_recall_macro        1.00000  1.00000  1.00000  1.00000  1.00000   
test_recall_weighted      0.98289  0.98289  0.98304  0.98304  0.98304   
train_recall_weighted     1.00000  1.00000  1.00000  1.00000  1.00000   
test_roc_auc              0.58521  0.63609  0.59919  0.60626  0.54997   
train_roc_auc             1.00000  1.00000  1.00000  1.00000  1.00000   

                          mean_score  
test_accuracy                0.98298  
train_accuracy               1.00000  
test_f1                      0.00000  
train_f1                     1.00000  
test_f1_macro                0.49571  
train_f1_macro               1.00000  
test_f1_weighted             0.97454  
train_f1_weighted            1.00000  
test_precision               0.00000  
train_precision              1.00000  
test_precision_macro         0.49149  
train_precision_macro        1.00000  
test_precision_weighted      0.96625  
train_precision_weighted     1.00000  
test_recall                  0.00000  
train_recall                 1.00000  
test_recall_macro            0.50000  
train_recall_macro           1.00000  
test_recall_weighted         0.98298  
train_recall_weighted        1.00000  
test_roc_auc                 0.59534  
train_roc_auc                1.00000

#### Key Takeaways
- Limiting feature set to the top features did not improve results at all with out of the box settings
- Let's try one more scenario using the hyperparameters from our grid search

#### Using `GridSearchCV` Hyperparameters and Top Features

In [61]:
rf_clf = RandomForestClassifier(n_estimators=200,
                                max_depth=2,
                                min_samples_split=2,
                                class_weight='balanced',
                                random_state=0,
                                n_jobs=-1,
                                verbose=2)

In [62]:
skf_results = cross_validate(rf_clf, # model
                             X_train_top, # feature matrix - X
                             y_train, # target vector - y
                             cv=skf, # cross-validation technique
                             scoring=scoring_metrics, # loss functions
                             return_train_score=True, # returns training score
                             verbose=2, # verbosity level to check progress of calculations
                             n_jobs=-1) # uses all computer cores              

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   14.9s finished


In [63]:
all_features_results = pd.DataFrame(skf_results).drop(['fit_time', 'score_time'], axis=1)
all_features_results_summary = pd.DataFrame(all_features_results.T)
all_features_results_summary['mean_score'] = all_features_results_summary.mean(axis=1)
all_features_results_summary

0        1        2        3        4  \
test_accuracy             0.66121  0.65682  0.65399  0.65853  0.65706   
train_accuracy            0.66391  0.65331  0.66038  0.66451  0.66184   
test_f1                   0.05544  0.05325  0.04750  0.05351  0.05329   
train_f1                  0.05626  0.05614  0.05800  0.05751  0.05747   
test_f1_macro             0.42451  0.42184  0.41805  0.42260  0.42195   
train_f1_macro            0.42590  0.42190  0.42542  0.42672  0.42571   
test_f1_weighted          0.78096  0.77782  0.77603  0.77916  0.77810   
train_f1_weighted         0.78298  0.77522  0.78033  0.78336  0.78141   
test_precision            0.02911  0.02794  0.02492  0.02807  0.02795   
train_precision           0.02954  0.02943  0.03044  0.03020  0.03017   
test_precision_macro      0.50911  0.50828  0.50608  0.50846  0.50839   
train_precision_macro     0.50949  0.50957  0.51021  0.50996  0.50997   
test_precision_weighted   0.97269  0.97217  0.97092  0.97256  0.97253   
train_precision_weighted  0.97312  0.97338  0.97363  0.97337  0.97343   
test_recall               0.58120  0.56410  0.50862  0.56897  0.56897   
train_recall              0.58925  0.60645  0.61373  0.60086  0.60515   
test_recall_macro         0.62190  0.61127  0.58256  0.61452  0.61377   
train_recall_macro        0.62723  0.63029  0.63746  0.63324  0.63399   
test_recall_weighted      0.66121  0.65682  0.65399  0.65853  0.65706   
train_recall_weighted     0.66391  0.65331  0.66038  0.66451  0.66184   
test_roc_auc              0.68885  0.63292  0.64106  0.65232  0.65403   
train_roc_auc             0.68232  0.68422  0.68891  0.68571  0.68904   

                          mean_score  
test_accuracy                0.65752  
train_accuracy               0.66079  
test_f1                      0.05260  
train_f1                     0.05708  
test_f1_macro                0.42179  
train_f1_macro               0.42513  
test_f1_weighted             0.77841  
train_f1_weighted            0.78066  
test_precision               0.02760  
train_precision              0.02996  
test_precision_macro         0.50807  
train_precision_macro        0.50984  
test_precision_weighted      0.97218  
train_precision_weighted     0.97338  
test_recall                  0.55837  
train_recall                 0.60309  
test_recall_macro            0.60880  
train_recall_macro           0.63244  
test_recall_weighted         0.65752  
train_recall_weighted        0.66079  
test_roc_auc                 0.65383  
train_roc_auc                0.68604

### Final Takeaways
- `RandomForestClassifier` results in similar results to `LogisticRegression` and `STS` model in terms of `roc_auc` scores